### PR TITLE
Changes for MDL-80677

### PR DESCRIPTION
**Pull request by andrewnicols**

### CODE REVIEW

**Review by junpataleta — COMMENTED** [review]
Thanks, Andrew! This looks good to me overall.

Should we also update the [Deprecation](https://moodledev.io/general/development/policies/deprecation) policy to provide information on using the `core/deprecated` attribute and `core/deprecation` during initial and final deprecation?

**Comment by andrewnicols** [comment]
I considered this but I don't know. It's not intended to replace the current deprecation, but to aid it. Maybe a note like:
```
:::note

From Moodle 4.4 onwards, you may wish to use the `\core\deprecated` attribute in combination with the `\core\deprecation` class to emit this message.

:::

**Comment by timhunt** [comment]
The buggest problem with Moodle deprectation messages is that people tell you to stop doing something without telling you what to do instead.

If what to do instead is not requred information, this is moving deckchairs on the Titanic. Certainly, all the give examples in the new docs should set a good example "reason: 'No longer required'," is so bad, it is basically a parody, right?

File: docs/apis/core/deprecation/index.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+// On a global function:
+#[\core\deprecated('random_bytes_emulate', since: '4.3', replacement: 'random_bytes')]
+function random_bytes_emulate($length) {}
+
+// On a class:
+#[\core\deprecated('\core\example', since: '4.4', reason: 'This functionality has been removed.')]
+class example {
+
+    #[\core\deprecated(
+        '\core\example::do_something',
```
[the remark this answers is not in the document]
Comment (andrewnicols) [reply docs/apis/core/deprecation/index.md:52]: Unfortunately it is not possible for an attribute to know what it belongs to without having something tell it, for example doing this in the code which loads the attribute:

```
            if ($attributes = $ref->getAttributes(deprecated::class)) {
                $attribute = $attributes[0]->newInstance();
                $attribute->set_owner($ref);
                return $attribute;
            }
```

That is not always possible.

**Comment by andrewnicols** [comment]
Thanks for the feedback Tim.

I thought about making this a required field, but there is not always a replacement. Sometimes things are just removed and that should be mentioned in the reason.

Do you have a better suggestion on how we can do this?

**Comment by timhunt** [comment]
When you say "I thought about making this a required field, but there is not always a replacement. Sometimes things are just removed and that should be mentioned in the reason." you are thinking about things from the wrong angle.

Think about things from the point of view as a plugin developer.

You don't just call Moodle core APIs for a laugh. Any time you are calling a core API, an you are trying to achieve some end goal on behalf of the user.

Generally, Moodle does not make it impossible to do things that were possible in the past. There must always be some sort of new recommended way to do thin thing that you were doing the old way.

Anyway, when you make an API change that broke my plugin, I don't really care why you thought it was a good idea. I just want to know what I should do now in my code. So, that should be the required information.

**Review by junpataleta — APPROVED** [review]

**Pull request merged**